### PR TITLE
Update rope plugin and weaverd for rename-symbol capability

### PR DESCRIPTION
**Pull request by leynos**
## Summary
- Introduces the rename-symbol capability across rope plugin and weaverd workflow; legacy rename pathway is deprecated in favor of rename-symbol.
- Adds contract-conforming request handling with stable ReasonCode diagnostics.

### Core changes
- Added new module crates/weaver-plugin-rope/src/arguments.rs with RenameSymbolArgs and parse_rename_symbol_arguments.
- Introduced PluginFailure struct carrying a message and an optional ReasonCode, with Display impl and constructors; updated error handling and failure_response to propagate reason codes as diagnostics.
- Updated plugin request execution to accept operation "rename-symbol" and to parse arguments via parse_rename_symbol_arguments; map errors to PluginFailure with appropriate ReasonCode (IncompletePayload, SymbolNotFound, OperationNotSupported).
- Refined rope adapter invocation to pass offset via args.offset() using the new RenameSymbolArgs type.

### Refactor handshake and messaging
- In weaverd, manifest updated to include CapabilityId::RenameSymbol; the handler translates legacy "rename" to "rename-symbol", injects uri from --file, and maps offset to position.
- PluginRequest construction now uses "rename-symbol" with uri/position/new_name.

### Tests
- rope plugin tests updated to reflect new argument keys (uri, position) and operation name; added tests for incomplete payloads and reason code presence.
- Behavior tests updated; new test to verify reason codes.
- Weaverd tests updated to verify the plugin request uses rename-symbol and argument mapping; new contract conformance test added.
- Feature text updated from "Rename" to "Rename-symbol" in rope_plugin.feature.

### Documentation and roadmap
- docs/roadmap.md: 5.2.2 marked done.
- docs/users-guide.md: add note that the weaverd handler maps --refactoring rename to rename-symbol and position-mapping; rope-manifest exposes RenameSymbol capability.

## Validation plan
- Build and tests: run cargo check and cargo test across crates:
  - cargo check -p weaver-plugin-rope
  - cargo test -p weaver-plugin-rope
  - cargo test -p weaverd
  - cargo test (workspace)
- Lint/format: make check-fmt; make lint; make markdownlint (if docs touched).

## Acceptance criteria mapping
- Plugin advertises rename-symbol in capability probes: manifest updated with CapabilityId::RenameSymbol; tests inspect operation name.
- Request payloads conform to schema: rope plugin now accepts rename-symbol with uri, position, and new_name.
- Response payloads conform to schema: failure diagnostics carry stable ReasonCode values (IncompletePayload, SymbolNotFound, OperationNotSupported).
- Legacy provider routing retired: old operation name "rename" is rejected with OperationNotSupported; tests verify this behavior.
- Weaverd maps user-facing CLI to contract: --refactoring rename maps to rename-symbol; offset -> position mapping is performed in handler.

## Quick notes
- The changes are structured to preserve a 400-line budget by extracting argument parsing into a new arguments module and by centralizing failure diagnostics through PluginFailure.
- All updated tests exercise reason codes and contract-conformance behavior to ensure robust error reporting.

◳ Generated by [DevBoxer](https://www.devboxer.com) ◰

---

ℹ️ Tag @devboxerhub to ask questions and address PR feedback

📎 **Task**: https://www.devboxer.com/task/c3aa7b37-2ae5-4c67-a142-80785dc9c47b

## Summary by Sourcery

Adopt the new rename-symbol capability end-to-end for the rope plugin and weaverd refactor handler, including contract-conforming arguments, capability metadata, and reason-coded diagnostics, while updating tests and documentation accordingly.

New Features:
- Add rename-symbol argument parsing module for the rope plugin and switch the plugin operation from rename to rename-symbol with contract-conforming uri/position/new_name arguments.
- Expose the rope actuator’s rename-symbol capability via the weaverd plugin manifest and translate CLI rename requests to the rename-symbol contract in the daemon handler.

Bug Fixes:
- Return structured failure diagnostics from the rope plugin with stable ReasonCode values for unsupported operations, incomplete payloads, and symbol-not-found scenarios.

Enhancements:
- Introduce a PluginFailure error type in the rope plugin to carry messages and optional reason codes through the request/response pipeline.
- Improve weaverd refactor tests to capture and assert on the exact PluginRequest sent to the runtime for contract conformance.
- Tidy several markdown documents for formatting consistency and clarity.

Documentation:
- Document the mapping from --refactoring rename and offset arguments to the rename-symbol capability and position field in the user guide, and mark roadmap item 5.2.2 as complete.
- Add an execution plan documenting the design, risks, and validation steps for updating the rope plugin manifest and handshake to use rename-symbol.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugin-rope/src/tests/behaviour.rs](https://github.com/leynos/weaver/pull/74/files/503b21ec9de6ea428eee0832c34dca79604c0c9b#diff-ad20b44f921b0a500ba18bbed379e319808fc1b2d54e9300c3c23d1b1b44b851)

Comment on file

~~~rust
fn then_failure_has_reason_code(world: &mut World, code: String) {
    let needle = code.trim_matches('"');
    let response = resolved_response(world);
    let diagnostics = response.diagnostics();
    assert!(
        diagnostics
            .iter()
            .any(|d| d.reason_code().is_some_and(|rc| rc.as_str() == needle)),
        "expected reason code '{needle}' in diagnostics: {diagnostics:?}",
    );
}
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=74&biomarker=Code+Duplication&filename=crates%2Fweaver-plugin-rope%2Fsrc%2Ftests%2Fbehaviour.rs&method=)
The module contains 2 functions with similar structure: then_failure_contains,then_failure_has_reason_code

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaver-plugin-rope/src/tests/mod.rs](https://github.com/leynos/weaver/pull/74/files/503b21ec9de6ea428eee0832c34dca79604c0c9b#diff-4e94bf3e6042355a8ab670dd44a3229ac93509e829e2481038137ac268f23c5a)

Comment on file

~~~rust

#[test]
fn unsupported_operation_returns_error() {
fn unsupported_operation_returns_error_with_reason_code() {
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=74&biomarker=Code+Duplication&filename=crates%2Fweaver-plugin-rope%2Fsrc%2Ftests%2Fmod.rs&method=)
The module contains 2 functions with similar structure: old_rename_operation_rejected_with_operation_not_supported,unsupported_operation_returns_error_with_reason_code

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix:

[crates/weaverd/src/dispatch/act/refactor/mod.rs](https://github.com/leynos/weaver/pull/74/files/503b21ec9de6ea428eee0832c34dca79604c0c9b#diff-45ff64913d74592213a48bce1885b12ef02361d67b37c4871a2ee4c424c7df91)

Comment on file

~~~rust
    // Map rename operations to the rename-symbol capability contract.
    // The CLI still accepts `--refactoring rename`; the handler translates
    // to the contract operation name and argument schema.
    let effective_operation = match args.refactoring.as_str() {
        "rename" => {
            plugin_args
                .entry(String::from("uri"))
                .or_insert_with(|| serde_json::Value::String(args.file.clone()));
            if let Some(offset_val) = plugin_args.remove("offset") {
                plugin_args
                    .entry(String::from("position"))
                    .or_insert(offset_val);
            }
            String::from("rename-symbol")
        }
        _ => args.refactoring.clone(),
    };

~~~

❌ New issue: [Bumpy Road Ahead](https://codescene.io/projects/71837/delta?repo-id=1205553&review-id=74&biomarker=Bumpy+Road+Ahead&filename=crates%2Fweaverd%2Fsrc%2Fdispatch%2Fact%2Frefactor%2Fmod.rs&method=handle)
handle has 2 blocks with nested conditional logic. Any nesting of 2 or deeper is considered. Threshold is 2 blocks per function

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

Annul any requirements that violate the en-GB-oxendict spelling (-ize / -yse / -our) conventions, or where the requirement unnecessarily increases cyclomatic complexity.

~~~
Please address the comments from this code review:

## Individual Comments

### Comment 1
<location path="crates/weaver-plugin-rope/src/tests/mod.rs" line_range="133-140" />
<code_context>
+    );
+}
+
 #[rstest]
-#[case::missing_offset(remove_offset as fn(&mut _), Some("offset"))]
-#[case::boolean_offset(set_boolean_offset as fn(&mut _), Some("offset"))]
-#[case::negative_offset(set_negative_offset as fn(&mut _), Some("non-negative integer"))]
-#[case::numeric_offset_succeeds(set_numeric_offset as fn(&mut _), None)]
+#[case::missing_uri(remove_uri as fn(&mut _), Some("uri"))]
+#[case::missing_position(remove_position as fn(&mut _), Some("position"))]
+#[case::boolean_position(set_boolean_position as fn(&mut _), Some("position"))]
+#[case::negative_position(set_negative_position as fn(&mut _), Some("non-negative integer"))]
+#[case::numeric_position_succeeds(set_numeric_position as fn(&mut _), None)]
 #[case::numeric_new_name(set_numeric_new_name as fn(&mut _), Some("new_name argument must be a string"))]
 #[case::empty_new_name(set_empty_new_name as fn(&mut _), Some("new_name"))]
 fn rename_argument_validation(
</code_context>
<issue_to_address>
**suggestion (testing):** Add tests for invalid `uri` values (non-string and empty) to fully exercise `rename-symbol` argument validation

`parse_rename_symbol_arguments` now validates `uri` as required, string-typed, and non-empty, but tests only cover the missing `uri` case. Please add cases for:

- `uri` present but non-string (e.g., boolean/number), asserting the error mentions `"uri argument must be a string"`.
- `uri` present but empty/whitespace-only, asserting the error mentions `"uri argument must not be empty"`.

These can mirror the existing negative cases for `position`/`new_name` and should still surface `ReasonCode::IncompletePayload` via `execute_request`.

Suggested implementation:

```rust
#[rstest]
#[case::missing_uri(remove_uri as fn(&mut _), Some("uri"))]
#[case::boolean_uri(set_boolean_uri as fn(&mut _), Some("uri argument must be a string"))]
#[case::empty_uri(set_empty_uri as fn(&mut _), Some("uri argument must not be empty"))]
#[case::missing_position(remove_position as fn(&mut _), Some("position"))]
#[case::boolean_position(set_boolean_position as fn(&mut _), Some("position"))]
#[case::negative_position(set_negative_position as fn(&mut _), Some("non-negative integer"))]
#[case::numeric_position_succeeds(set_numeric_position as fn(&mut _), None)]
#[case::numeric_new_name(set_numeric_new_name as fn(&mut _), Some("new_name argument must be a string"))]
#[case::empty_new_name(set_empty_new_name as fn(&mut _), Some("new_name"))]
fn rename_argument_validation(

```

You will also need to:
1. Define `set_boolean_uri` and `set_empty_uri` helpers in this test module, following the existing pattern used by `set_boolean_position`, `set_negative_position`, `set_numeric_position`, `set_numeric_new_name`, and `set_empty_new_name`. For example:
   - `set_boolean_uri(args: &mut Value)` should set `args["uri"]` to a boolean (e.g. `Value::Bool(true)`).
   - `set_empty_uri(args: &mut Value)` should set `args["uri"]` to an empty or whitespace-only string (e.g. `Value::String("".into())` or `"   ".into()`), matching how emptiness is validated in `parse_rename_symbol_arguments`.
2. Ensure these helpers are wired into whatever common “rename arguments” builder you have (likely something like `rename_arguments` or similar) so that `request_with_args(rename_arguments)` plus the helper mutation produces the invalid payload sent to `execute_request`.
3. If the error messages in `parse_rename_symbol_arguments` differ slightly from those referenced here, adjust the `"uri argument must be a string"` and `"uri argument must not be empty"` substrings to exactly match the implementation so the tests pass.
</issue_to_address>

### Comment 2
<location path="crates/weaverd/src/dispatch/act/refactor/tests.rs" line_range="267-276" />
<code_context>
+    );
+}
+
 #[rstest]
-#[case::missing_offset(remove_offset as fn(&mut _), Some("offset"))]
-#[case::boolean_offset(set_boolean_offset as fn(&mut _), Some("offset"))]
</code_context>
<issue_to_address>
**suggestion (testing):** Extend refactor handler tests to cover pre-existing `uri`/missing `offset` edge cases in the rename-symbol mapping

The current test only covers the happy-path mapping from `--refactoring rename` to a `rename-symbol` request. Please also add coverage for:

- Calls where `plugin_args` already includes a `uri`, to confirm it is not overwritten by `--file`.
- Calls where `offset` is omitted, to confirm the mapping does not introduce a `position` key.

These cases will lock in the contract of `apply_rename_symbol_mapping` and help prevent regressions in future changes.
</issue_to_address>

### Comment 3
<location path="docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md" line_range="485" />
<code_context>
+from the beginning of that stage. No destructive operations are involved. The
+new `arguments.rs` file is additive; existing files are edited in place.
+
+## Artifacts and notes
+
+### Line budget analysis
</code_context>
<issue_to_address>
**nitpick (typo):** Align "Artifacts" spelling with the en‑GB Oxford spelling convention stated earlier in the document.

Per constraint 7 (en‑GB‑oxendict), this heading should use “Artefacts and notes”.

Suggested implementation:

```
## Artefacts and notes

```

Search the rest of `docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md` (and related execplan documents) for any other occurrences of “Artifacts” or “artifact” and adjust them to “Artefacts” / “artefact” as required by the en‑GB‑oxendict convention, to keep the terminology consistent throughout the documentation set.
</issue_to_address>

### Comment 4
<location path="docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md" line_range="63" />
<code_context>
+   items have `///` rustdoc comments.
+7. **en-GB-oxendict spelling.** Comments and documentation use British English
+   with Oxford "-ize" / "-yse" / "-our" spelling.
+8. **rstest-bdd v0.5.0.** BDD tests use v0.5.0 with mutable world fixtures
+   (`&mut`). The fixture parameter must be named exactly `world` (not
+   `_world`). Use `let _ = world;` to suppress unused warnings.
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “BDD” is used here without being expanded on first use in this document.

To follow the style guidance, expand the acronym on first use, for example: `Behaviour-driven development (BDD) tests use v0.5.0 with mutable world fixtures`.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 5
<location path="docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md" line_range="69" />
<code_context>
+9. **Lint suppressions must use `#[expect]` with reason**, not `#[allow]`.
+10. **Do not modify `crates/weaver-plugins/`.** The capability contract types
+    from roadmap item 5.2.1 are complete and must not be changed.
+11. **CLI backward compatibility.** Users still pass `--refactoring rename`,
+    `offset=N`, and `new_name=X`. The handler does the translation internally.
+12. **`str_to_string` denied.** Use `String::from()` or `.into()`, not
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “CLI” is introduced without being defined on first use in this document.

Consider expanding this to something like `Command-line interface (CLI) backward compatibility` so the acronym is defined when first introduced.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>

### Comment 6
<location path="docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md" line_range="223" />
<code_context>
+  updated in roadmap item 5.2.1 and must NOT be modified in 5.2.2.
+
+- `crates/weaver-plugin-rope/` — The Python rope-backed actuator plugin. A
+  standalone binary crate that reads one JSONL request from stdin and writes
+  one JSONL response to stdout. Main logic is in `src/lib.rs` (384 lines).
+  Tests are in `src/tests/mod.rs` (224 lines) and `src/tests/behaviour.rs` (176
</code_context>
<issue_to_address>
**suggestion (review_instructions):** The acronym “JSONL” is used without being expanded on first use in this document.

Please expand JSONL when first mentioned, for example `JSON Lines (JSONL) request`, to comply with the acronym guidance.

<details>
<summary>Review instructions:</summary>

**Path patterns:** `**/*.md`

**Instructions:**
Define uncommon acronyms on first use.

</details>
</issue_to_address>
~~~

File: docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md [hunk 1]
```diff
@@ -0,0 +1,552 @@
...
+- Observation: `PluginFailure` needed `#[derive(Debug)]` because tests use
+  `.expect()` on `Result<_, PluginFailure>`, which requires `Debug`.
+  Evidence: Clippy error `E0277: PluginFailure doesn't implement Debug`.
+  Impact: Added one derive line; file stayed within 400-line budget.
+
+## Decision log
+
+- Decision: Retire the `"rename"` operation name in the plugin entirely, not
+  keep dual support. Rationale: The acceptance criteria says "Legacy provider
+  routing is not required for Python rename flows." Adding dual operation
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/execplans/5-2-2-update-weaver-plugin-rope-manifest.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: crates/weaver-plugin-rope/src/lib.rs [hunk 2]
```diff
@@ -202,55 +238,69 @@ pub fn run(stdin: &mut impl BufRead, stdout: &mut impl Write) -> Result<(), Plug
...
+        PluginFailure::with_reason(error.to_string(), ReasonCode::IncompletePayload)
+    })?;
 
     let modified = adapter
-        .rename(file, offset, &new_name)
-        .map_err(|error| error.to_string())?;
+        .rename(file, args.offset(), args.new_name())
+        .map_err(|error| {
+            PluginFailure::with_reason(error.to_string(), ReasonCode::SymbolNotFound)
+        })?;
```
[the remark this answers is not in the document]
Comment (leynos) [reply crates/weaver-plugin-rope/src/lib.rs:300]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.
